homeassistant/components/abode/config_flow.py

```diff
@@ -0,0 +1,66 @@
+"""Config flow for the Abode Security System component."""
+import logging
+
+import voluptuous as vol
+
+from abodepy.exceptions import AbodeAuthenticationException
+from homeassistant import config_entries
+
+from homeassistant.const import CONF_PASSWORD, CONF_USERNAME
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@config_entries.HANDLERS.register(DOMAIN)
+class AbodeFlowHandler(config_entries.ConfigFlow):
+    """Config flow for Abode."""
+
+    VERSION = 1
+
+    def __init__(self):
+        """Initialize."""
+        self.data_schema = {}
+        self.data_schema[vol.Required(CONF_USERNAME)] = str
+        self.data_schema[vol.Required(CONF_PASSWORD)] = str
+
+    async def async_step_user(self, user_input=None):
+        """Handle a flow initialized by the user."""
+        from abodepy import Abode
+
+        if self._async_current_entries():
+            return self.async_abort(reason="single_instance_allowed")
+
+        if not user_input:
+            return await self._show_form()
+
+        username = user_input[CONF_USERNAME]
+        password = user_input[CONF_PASSWORD]
+
+        try:
+            self.hass.async_add_executor_job(Abode, username, password, True)
+
+        # Need to add error checking if Abode server is down/not responding
```

Comment:
This remains to do.

Reply:
I think this will require updates to the abodepy module. Right now, this is how `response.status_code` other than 200 is handled:

```
if response.status_code != 200:
            raise AbodeAuthenticationException((response.status_code,
                                                response_object['message']))
```

All that we get back in Home Assistant is a class object, `<class 'abodepy.exceptions.AbodeAuthenticationException'>`.

If I'm understanding this correctly, we would need abodepy to throw a different type of exception for the server being down that we could then use in the Abode component in HA to create another type of error to show the user.

Reply:
So it appears abodepy does pass the error code and error message in the exception, so I should be able to differentiate between invalid credentials (which is being passed as response code 400) and the server not responding, which I need to figure out.

Reply:
I'm not sure how to figure out what the response code is for the abode server being down. I'm sure there's a few different cases we would have to cover. The only thing I can really is simulate is your own ISP being down which HA already throws an error for. However, I did make some adjustments to make the error catching more explicit for the credentials being invalid, then sort of a catch all for any other errors on the Abode server side called "abode_error" with a message "Error occurred with Abode servers."
```
        except AbodeAuthenticationException as error:
            if error.errcode == 400:
                return await self._show_form({"base": "invalid_credentials"})
            return await self._show_form({"base": "abode_error"})
```
I think this implementation covers what we need and at least provides the user with more descriptive feedback.